src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -639,257 +659,209 @@ void xocl_drm_fini(struct xocl_drm *drm_p)
+static int xocl_mm_insert_node_range_all(struct xocl_drm *drm_p, uint32_t *mem_id,
 		struct mem_topology *grp_topology, struct drm_mm_node *dnode, u64 size)
 {
-	struct xocl_mm_wrapper *wrapper = NULL;
+	struct mem_data *mem_data = NULL;
+	struct xocl_mm *xocl_mm = drm_p->xocl_mm;
 	uint64_t start_addr = 0;
 	uint64_t end_addr = 0;
-	uint64_t hash_start_addr = 0;
 	int ret = 0;
 	int i = 0;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 7, 0)
+	BUG_ON(!xocl_mm && !xocl_mm->mm);
+
 	for (i = 0; i < grp_topology->m_count; i++) {
-		if ((convert_mem_tag(grp_topology->m_mem_data[i].m_tag) == MEM_TAG_HOST) ||
-				XOCL_IS_PS_KERNEL_MEM(grp_topology, i))
+		mem_data = &grp_topology->m_mem_data[i];
+		if ((convert_mem_tag(mem_data->m_tag) == MEM_TAG_HOST) ||
+				!XOCL_IS_PS_KERNEL_MEM(grp_topology, i))
 			continue;
 
-		hash_start_addr = grp_topology->m_mem_data[i].m_base_address;
-		hash_for_each_possible(drm_p->mm_range, wrapper, node, hash_start_addr) {
-			start_addr = wrapper->start_addr;
-			end_addr = wrapper->start_addr + wrapper->size;
+		start_addr = mem_data->m_base_address;
+		end_addr = start_addr + mem_data->m_size;
 
 #if defined(XOCL_DRM_FREE_MALLOC)
-			ret = drm_mm_insert_node_in_range(drm_p->mm, dnode, size, PAGE_SIZE, 0,
-					start_addr, end_addr, 0);
+		ret = drm_mm_insert_node_in_range(xocl_mm->mm, dnode, size, PAGE_SIZE, 0,
+				start_addr, end_addr, 0);
 #else
-			ret = drm_mm_insert_node_in_range(drm_p->mm, dnode, size, PAGE_SIZE,
-					start_addr, end_addr, 0);
+		ret = drm_mm_insert_node_in_range(xocl_mm->mm, dnode, size, PAGE_SIZE,
+				start_addr, end_addr, 0);
 #endif
-
-			if (!ret) 
-				/* Found the memory. Return it from here  */
-				return 0;
+		if (!ret) {
+			// Memory is allocated to this Bank
+			*mem_id = i;
+			return 0;
 		}
-	}	
-#endif
+	}
 
 	return ret;
 }
 
-
 static int xocl_mm_insert_node_range(struct xocl_drm *drm_p, u32 mem_id,
 		struct mem_topology *grp_topology, struct drm_mm_node *node, u64 size)
 {
+	struct xocl_mm *xocl_mm = drm_p->xocl_mm;
 	uint64_t start_addr = 0;
 	uint64_t end_addr = 0;
 	int ret = 0;
 
+	BUG_ON(!xocl_mm && !xocl_mm->mm);
 	start_addr = grp_topology->m_mem_data[mem_id].m_base_address;
 	end_addr = start_addr + grp_topology->m_mem_data[mem_id].m_size * 1024;
 
 #if defined(XOCL_DRM_FREE_MALLOC)
-	ret = drm_mm_insert_node_in_range(drm_p->mm, node, size, PAGE_SIZE, 0,
+	ret = drm_mm_insert_node_in_range(xocl_mm->mm, node, size, PAGE_SIZE, 0,
 					   start_addr, end_addr, 0);
 #else
-	ret = drm_mm_insert_node_in_range(drm_p->mm, node, size, PAGE_SIZE,
+	ret = drm_mm_insert_node_in_range(xocl_mm->mm, node, size, PAGE_SIZE,
 					   start_addr, end_addr, 0);
 #endif
 
 	return ret;
 }
 
-int xocl_mm_insert_node(struct xocl_drm *drm_p, u32 mem_id,
-			struct drm_mm_node *node, u64 size)
+int xocl_mm_insert_node(struct xocl_drm *drm_p, unsigned memidx,
+			uint32_t slotidx, struct drm_mm_node *node, u64 size)
```

Comment:
This is confusing. The slotidx is per xclbin, correct? It is never used with xocl_mm_insert_*. Is it possible that bo carries the incorrect slotidx which belongs to other xclbin?

Reply:
The slot id is required for status update. Along with the device memory stats we are also maintaining the stats for each slot.